### ISSUE TITLE
Document flexmark dependency for test reports

### ISSUE THREAD

**Issue by sloshy** [scalatest/scalatest#1736]
Currently ScalaTest does not seem to indicate anywhere that it now needs flexmark as a dependency for doing test reports. In addition, the only supported version of flexmark (at the time) appears to be `0.35.10`. This should be clearly documented somewhere so as to not trip anybody up expecting to be able to use the latest version.

**Comment by peterzeller** [comment]
Where exactly do I have to add this dependency in an SBT project? 

I tried adding `libraryDependencies += "com.vladsch.flexmark" % "flexmark-all" % "0.35.10"` to `build.sbt` and to `project/build.sbt`, but I still  get `java.lang.NoClassDefFoundError: com/vladsch/flexmark/ast/Node`.

**Comment by jan-pieter** [comment]
@peterzeller I encountered the same issue. Downgrading sbt worked for me. See https://github.com/sbt/sbt/issues/5308

**Comment by cstroe** [comment]
I have the same issue with my Gradle build and I added the following in my `dependencies` block:

```
testRuntimeOnly "com.vladsch.flexmark:flexmark-all:0.35.10"
```

**Comment by liebichw** [comment]
I also fell for it  and mistakenly took that for a bug in sbt.
To demonstrate the issue I created a minimal demonstration project in https://github.com/liebichw/scalatest-bug-demonstration.

**Comment by nafg** [comment]
Perhaps someone wants to open a PR to fix this?

**Comment by DanySK** [comment]
I was able to limit the imported flexmark dependencies. In Gradle 6.3:
`testRuntimeOnly("com.vladsch.flexmark:flexmark-profile-pegdown:0.36.8")`

Later versions (0.40.x and above fail with):

```
An exception or error caused a run to abort. This may have been caused by a problematic custom reporter.
java.lang.NoClassDefFoundError: com/vladsch/flexmark/ast/Node
        at org.scalatest.tools.ReporterFactory.createHtmlReporter(ReporterFactory.scala:194)
```

Besides the need for documentation, I wonder if it makes sense to use/support more recent versions of flexmark.

**Comment by liebichw** [comment]
What kind of dependency should I declare in a build.sbt file? A "test" dependency? BTW, flexmark-all 0.35.10 also pulls in a dependency on log4j 1.2.17  - my project uses already log4j2, so getting this additional log4j version brings in some strange problems.
flexmark-all 0.36.8 seems to fix that problem.

**Comment by nafg** [comment]
Is this going to be addressed?

**Comment by bejelith** [comment]
I started having the same problem with gradle upgrading from 3.0 to 3.1 and 3.2

**Comment by cheeseng** [comment]
Hi, fyi I submitted a PR to update flexmark version to 0.62.2: 

https://github.com/scalatest/scalatest/pull/2050

**Comment by DanySK** [comment]
Nice! Upvoted already

**Comment by caoilte** [comment]
@cheeseng I wonder if it would be worth updating it to be non-optional at the same time. Otherwise everyone who upgrades to the next version of scalatest is going to have to re-apply the workaround in this ticket for the lastest version. 

I can't see any explanation for why the dependency was originally marked optional. Perhaps at some point in the history of sbt html reports weren't turned on by default. I can't think of any other reason that makes sense.

**Comment by cheeseng** [comment]
@caoilte The reason is not everybody use HtmlReporter, those who doesn't use it probably won't like to pull the extra dependency in, we're trying to limit non-optional dependency.  The other way may be we should be clearer on the documentation about the need to include the dependency when using HtmlReporter though.

**Comment by DanySK** [comment]
@cheeseng one way I invite you to consider is to *fail very clearly*: when the HtmlReporter is enabled but flexmark is not there, fail with a message that states something like:
```
In order to enable HTML reporting in scalatest, the user should add the following dependency to the test runtime:
        group: com.vladsch.flexmark
        module: flexmark-profile-pegdown
        version: 0.36.8

Gradle users can do so, for instance, through:
        dependencies { testRuntimeOnly("com.vladsch.flexmark:flexmark-profile-pegdown:0.36.8") }`

sbt users can do so through:
        <how it is done in sbt>
```

That would in my opinion partially compensate for the lack of documentation

**Comment by cheeseng** [comment]
> @cheeseng one way I invite you to consider is to _fail very clearly_: when the HtmlReporter is enabled but flexmark is not there, fail with a message that states something like:
> 
> ```
> In order to enable HTML reporting in scalatest, the user should add the following dependency to the test runtime:
>         group: com.vladsch.flexmark
>         module: flexmark-profile-pegdown
>         version: 0.36.8
> 
> Gradle users can do so, for instance, through:
>         dependencies { testRuntimeOnly("com.vladsch.flexmark:flexmark-profile-pegdown:0.36.8") }`
> 
> sbt users can do so through:
>         <how it is done in sbt>
> ```
> 
> That would in my opinion partially compensate for the lack of documentation

@DanySK Sounds like a good idea, let's try to do that.

**Comment by DanySK** [comment]
@cheeseng also please consider merging #2050 before addressing the issue, so that we end up with recommending a modern version of Flexmark.

**Comment by helfper** [comment]
Why not have batteries included by default? People who care about having lean dependencies and don't need Flexmark could always exclude it when adding ScalaTest. And this would avoid people who do need it to have to check ScalaTest dependencies to figure out which version of Flexmark they should use.

**Comment by DanySK** [comment]
@helfper I agree with you, but if the policy is to not have possibly unused transitive dependencies, then it makes sense to at least fail clearly as second best solution.

**Comment by cheeseng** [comment]
@helfper @DanySK @caoilte Some background, we used to have other optional dependencies like scalacheck, junit, testng, mockito etc. as optional dependencies, as there were complains that bringing in scalatest will transitively bringing all these libs also.  That's where the optional policy came from.  Today, those integration parts are now separated into their own module like scalatest+scalacheck, scalatest+junit, scalatest+testng etc. so the only left 'optional' dependency now seems like flexmark.  One option may be is to have HtmlReporter separated out as scalatest-htmlreporter and they we can have the dependency set to non-optional, just like the other scalatest+scalacheck, scalatest+junit etc. does currently.

**Comment by cheeseng** [comment]
> @cheeseng also please consider merging #2050 before addressing the issue, so that we end up with recommending a modern version of Flexmark.

@DanySK  @bvenners will have the pleasure to review and merge it, I can attempt the error message enhancement in a separate PR.

**Comment by caoilte** [comment]
@cheeseng thanks for the clarification and I totally understand the motivation. IMO It would make more sense to make flexmark optional if the html reporter was turned off by default. That way the missing dependency would be an expected outcome from enabling the feature.

Also - is it even documented how to turn off the html reporter? I can't find out how to do it!

**Comment by cheeseng** [comment]
@caoilte Actually to use html reporter you have to add -h <target_dir> to the test options, -h as mentioned in the table here: 

https://www.scalatest.org/user_guide/using_the_runner

By default it is not turned on, it only will be turned on the -h is passed to scalatest.

**Comment by caoilte** [comment]
My bad. We have a "clever" plugin at work.

**Comment by DanySK** [comment]
@caoilte it was enabled by default with the gradle plugin for me. I disabled it this way: https://github.com/AlchemistSimulator/Alchemist/commit/a99dadb0ec038997aed1a953998ba32301dd40cb

**Comment by blakehawkins** [comment]
Is there a way to resolve this with libraryDependencies when using scalatest 3.1.4? I have tried pinning flexmark to 0.35.10 or 0.36.8 or 0.62.2 but I always get the same stacktrace

**Comment by gauravbrills** [comment]
`"com.vladsch.flexmark:flexmark-profile-pegdown:0.62.2"` worked for me on gradle-6.9 and scalaTestVersion=3.2.13
